### DISCUSSION TITLE
Services WithHostname and WithServiceBinding don't play together across modules

### DISCUSSION THREAD

**Discussion by marcosnils** [dagger/dagger#12197] [Help]
We should find a way to consolidate the API between `WithHostname` and `WithServiceBinding` as they currently don't play along very well together in a multi-module setup.

Repro:

Module A: 

```go
// Returns a container that echoes whatever string argument is provided
func (m *Lala) Test(ctx context.Context) *dagger.Container {

    svc := dag.Container().From("nginx").
        WithExposedPort(80).AsService().WithHostname("web")

    svc.Start(ctx)

    return dag.B().Run(svc)

}
```

Module B: 
```go
func (m *B) Run(svc *dagger.Service) *dagger.Container {

    return dag.Container().From("alpine").
        WithEnvVariable("CACHE", time.Now().String()).
        WithServiceBinding("foo", svc).
        WithExec([]string{"wget", "-O", "-", "http://foo"})
}
```

output: 
```
Error: response from query: input: container.from.withEnvVariable.withServlve: process "wget -O - http://foo" did not complete successfully: lookup on 10.87.0.1:53: no such host
lookup web.g4jodv4eddahq.iap51blr32pb4.dagger.local on 10.87.0.1:53: no su
lookup web.iap51blr32pb4.dagger.local on 10.87.0.1:53: no such host
```

error comes from here: https://github.com/marcosnils/dagger/blob/bdce0aa73eb7114bf48180e6c1edd9f59bdb0be4/engine/buildkit/executor_spec.go?plain=1#L275

cc @vito

**Comment by vito** [comment]
The catch is that we can only set a single `Hostname` with Buildkit for a service, so when `WithHostname` is used we set the configured one instead of the content-addressed one. That breaks in this case, since only one of the modules will have actually started it, and it'll only have that module's ID in the hostname.

Effectively, only one module can use a `WithHostname`d service at a time within a session. You would have the same problem if you tried to reach the configured hostname within another module.

So it's sort of fundamentally weird: services are keyed within a session by a cross-module value, but they only listen on a single module's address.

It's enough for some cases, but doesn't feel quite right.

Maybe it would be less bad if `WithHostname` baked in the calling module's domain? Then you could at least pass it around. Still feels hacky though.

I think ideally we would start a single service with the regular content-addressed hostname, and then for each module that starts the `WithHostname` service, they actually just install an additional DNS route for it somehow.
